### PR TITLE
BUG: 1D slices over extension types turn into N-dimensional slices over ExtensionArrays

### LINKED ISSUE #42430: BUG: 1D slices over extension types turn into N-dimensional slices over ExtensionArrays
- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the latest version of pandas.

- [X] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
class LoggingStringArray(pd.arrays.StringArray):
    """Extend StringArray to log arguments to __getitem__"""
    def __getitem__(self, item):
        print(f"Argument to __getitem__ is '{item}'")
        return super().__getitem__(item)

df = pd.DataFrame({"col1": LoggingStringArray(np.array(["hello", "world"], dtype=object))})
_ = df.iloc[:1]
```

Expected output:
```
Argument to __getitem__ is 'slice(None, 1, None)'
```

Actual output:
```
Argument to __getitem__ is '(Ellipsis, slice(None, 1, None))'
```

#### Problem description

If a DataFrame contains extension types, the block manager turns a 1D slice in the form `DataFrame.iloc[slice(x, y, z)]` into an N-dimensional slice over the ExtensionArray of the form `__getitem__((Ellipsis, slice(x, y, z))` . This problem breaks the `__getitem__()` method on one of our extension types in [Text Extensions for Pandas](https://github.com/CODAIT/text-extensions-for-pandas).

The `__getitem__` methods in Pandas' built-in extension types happens to work when fed this kind of n-dimensional slice
--- they just ignore the `Ellipsis` in the first position of the tuple. But I don't think they're supposed to accept that type of input.

The root cause is the new method `Block.getitem_block_index()`, added during some performance optimizations introduced in pandas-dev/pandas#40353:

```python
    def getitem_block_index(self, slicer: slice) -> Block:
        """
        Perform __getitem__-like specialized to slicing along index.

        Assumes self.ndim == 2
        """
        # error: Invalid index type "Tuple[ellipsis, slice]" for
        # "Union[ndarray, ExtensionArray]"; expected type "Union[int, slice, ndarray]"
        new_values = self.values[..., slicer]  # type: ignore[index]
        return type(self)(new_values, self._mgr_locs, ndim=self.ndim)
```

This method assumes that the contents of the block are 2-dimensional. The `ExtensionBlock` class inherits this method from `EABackedBlock`, which inherits it from `Block`.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : f00ed8f47020034e752baf0250483053340971b0
python           : 3.7.10.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 20.5.0
Version          : Darwin Kernel Version 20.5.0: Sat May  8 05:10:33 PDT 2021; root:xnu-7195.121.3~9/RELEASE_X86_64
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.3.0
numpy            : 1.21.0
pytz             : 2021.1
dateutil         : 2.8.1
pip              : 21.1.3
setuptools       : 52.0.0.post20210125
Cython           : None
pytest           : 6.2.4
hypothesis       : None
sphinx           : 4.0.2
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 3.0.1
IPython          : 7.25.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : 3.4.2
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 4.0.1
pyxlsb           : None
s3fs             : None
scipy            : 1.7.0
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : None

</details>

### PR DESCRIPTION
**Pull request by frreiss**
- [X] closes #42430
- [X] tests added / passed
- [X] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [X] whatsnew entry

This PR contains a 1-line patch for issue #42430, plus a regression test to ensure the issue does not recur.

Before this change, `ExtensionBlock. getitem_block_index()` would request 2D slices of a block's underlying extension array, even though the `__getitem__()` method of `ExtensionArray` is (currently) only supposed to support 1D slicing. After the change, `getitem_block_index()` requests 1D slices. The change also fixes some linter/type checking errors.

### CODE REVIEW

File: pandas/core/internals/blocks.py [hunk 1]
```diff
@@ -1552,12 +1552,8 @@ def _slice(self, slicer):
     def getitem_block_index(self, slicer: slice) -> ExtensionBlock:
         """
         Perform __getitem__-like specialized to slicing along index.
-
-        Assumes self.ndim == 2
         """
-        # error: Invalid index type "Tuple[ellipsis, slice]" for
-        # "Union[ndarray, ExtensionArray]"; expected type "Union[int, slice, ndarray]"
-        new_values = self.values[..., slicer]  # type: ignore[index]
+        new_values = self.array_values[slicer]
```
Comment (jbrockmendel) [inline pandas/core/internals/blocks.py:1556]: just .values, not .array_values

Comment (jbrockmendel) [reply pandas/core/internals/blocks.py:1556]: can you add a comment to the effect of `# GH#42787 in principle equivalent to values[..., slicer] but we don't require subclasses to support that (for now)`

File: pandas/tests/arrays/test_array.py [hunk 2]
```diff
@@ -317,6 +317,26 @@ def test_bounds_check():
         pd.array([-1, 2, 3], dtype="UInt16")
 
 
+def test_ellipsis_index():
+    # GH42430
```
Comment (jbrockmendel) [inline pandas/tests/arrays/test_array.py:321]: this would go in tests.extension.base.getitem

File: pandas/core/internals/blocks.py (hunk 1 above)
Comment (frreiss) [reply pandas/core/internals/blocks.py]: Changed to `.values` and added an explanatory comment.

File: pandas/tests/arrays/test_array.py (hunk 2 above)
Comment (frreiss) [reply pandas/tests/arrays/test_array.py]: Moved to `tests/extension/base/getitem.py`

**Comment by frreiss** [comment]
Looks like there was a single test failure on one of the CI builds. I'm not able to reproduce that failure on my machine. Can someone with the appropriate permissions rerun the `pytest (actions-39-numpydev.yaml, not slow and not network, xsel, deprecate, -W error)` build to see if the problem goes away by itself?

**Comment by jbrockmendel** [comment]
that should be fixed if you merge master

**Comment by jreback** [comment]
thanks @frreiss for the patch!

**Comment by jreback** [comment]
@meeseeksdev backport 1.3.x

**Comment by jreback** [comment]
@frreiss if you wouldn't mind following the instructions above to manually backport this (above).

**Comment by simonjayhawkins** [comment]
conflicts due to #42422. manual backport done in #42850

**Pull request merged**